### PR TITLE
add std to tensor.py

### PR DESCRIPTION
**Pull request by rabiaedayilmaz**

### CODE REVIEW

**Comment by geohot** [comment]
Needs tests, and at least one line too many.

**Comment by rabiaedayilmaz** [comment]
> Needs tests, and at least one line too many.

Test added and reduced to a one-liner.

**Comment by jla524** [comment]
I ran the following tests and they failed due to shape mismatch. Should the shapes match?
```
def test_std_axis(self):
  # AssertionError: shape mismatch (tinygrad)(1,) != (torch)(65, 85)
  helper_test_op([(45,65,85)], lambda x: torch.std(x,axis=0), lambda x: Tensor.std(x,axis=0))

def test_std_keepdim(self):
  # AssertionError: shape mismatch (tinygrad)(1,) != (torch)(1, 1, 1)
  helper_test_op([(45,65,85)], lambda x: torch.std(x,keepdim=True), lambda x: Tensor.std(x,keepdim=True))
```

**Comment by rabiaedayilmaz** [comment]
> I ran the following tests and they failed due to shape mismatch. Should the shapes match?
> 
> ```
> def test_std_axis(self):
>   # AssertionError: shape mismatch (tinygrad)(1,) != (torch)(65, 85)
>   helper_test_op([(45,65,85)], lambda x: torch.std(x,axis=0), lambda x: Tensor.std(x,axis=0))
> 
> def test_std_keepdim(self):
>   # AssertionError: shape mismatch (tinygrad)(1,) != (torch)(1, 1, 1)
>   helper_test_op([(45,65,85)], lambda x: torch.std(x,keepdim=True), lambda x: Tensor.std(x,keepdim=True))
> ```

Right, something odd, will be inspected.

**Comment by rabiaedayilmaz** [comment]
@jla524, I added the tests you provided, thank you. Now it passes those tests, no shape mismatch error anymore. However, in the test_std_axis function, I needed to change rtol to 1e-1, the values do not match exactly.

**Comment by rabiaedayilmaz** [comment]
I found the problem!! According to [1], Pytorch uses Bessel's correction to measure std which is different from Numpy std. To implement like Numpy, setting `unbiased=False`, fixes everything.

[1] https://stackoverflow.com/questions/69264984/why-does-numpy-and-pytorch-give-different-results-after-mean-and-variance-normal

**Comment by jla524** [comment]
> I found the problem!! According to [1], Pytorch uses Bessel's correction to measure std which is different from Numpy std. To implement like Numpy, setting `unbiased=False`, fixes everything.
> 
> [1] https://stackoverflow.com/questions/69264984/why-does-numpy-and-pytorch-give-different-results-after-mean-and-variance-normal

Excellent!

**Comment by geohot** [comment]
Hmm, very close. But it doesn't behave like torch, can you add the unbiased option with the same default? It doesn't look hard.
* (((n/(n-1)) if unbiased else 1)

**Comment by rabiaedayilmaz** [comment]
Good idea, added an unbiased option and it seems it works in both options now.

**Comment by geohot** [comment]
Test both. And there's def cleaner ways to write that if and reuse state. Line count is less important than cleanliness.

**Comment by rabiaedayilmaz** [comment]
Thank you for bearing with me. I have done them, tested both, and more clean code.

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -308,7 +308,9 @@ def mean(self, axis=None, keepdim=False):
     return out * (prod(out.shape)/prod(self.shape))
   def std(self, axis=None, keepdim=False, unbiased=True):
     n = self.shape[axis] if axis is not None else prod(self.shape)
-    return ((((self - self.mean(axis=axis, keepdim=True)).square()).sum(axis=axis, keepdim=keepdim))/(n-1)).sqrt() if unbiased else (((self - self.mean(axis=axis, keepdim=True)).square()).mean(axis=axis, keepdim=keepdim)).sqrt()
+    if unbiased:
+      return ((((self - self.mean(axis=axis, keepdim=True)).square()).sum(axis=axis, keepdim=keepdim)) / (n - 1)).sqrt()
+    return (((self - self.mean(axis=axis, keepdim=True)).square()).mean(axis=axis, keepdim=keepdim)).sqrt()
```
Comment (geohot) [inline tinygrad/tensor.py:313]: ` ((((self - self.mean(axis=axis, keepdim=True)).square()).sum(axis=axis, keepdim=keepdim)) / ((n - 1) if unbiased else n)).sqrt()

Comment (geohot) [reply tinygrad/tensor.py:313]: But even better

```
square_sum = (self - self.mean(axis=axis, keepdim=True)).square()).sum(axis=axis, keepdim=keepdim)
return (square_sum / (prod(square_sum.shape)-(1 if unbiased else 0)).sqrt()
```

Comment (rabiaedayilmaz) [reply tinygrad/tensor.py]: Fantastic, edited.

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -306,7 +306,10 @@ def min(self, axis=None, keepdim=False): return -((-self).max(axis=axis, keepdim
   def mean(self, axis=None, keepdim=False):
     out = self.sum(axis=axis, keepdim=keepdim)
     return out * (prod(out.shape)/prod(self.shape))
-
+  def std(self, axis=None, keepdim=False, unbiased=True):
+    n = self.shape[axis] if axis is not None else prod(self.shape)
```
Comment (geohot) [inline tinygrad/tensor.py:310]: Not with this!

Comment (rabiaedayilmaz) [reply tinygrad/tensor.py:310]: But `prod(square_sum.shape)` can not pass test_std_axis with "not equal tolerance", however above can pass.

Comment (rabiaedayilmaz) [reply tinygrad/tensor.py:310]: This passes all the tests. Is it eligible? `(square_sum / ((prod(self.shape[:axis+1]) if axis is not None else prod(self.shape))-(1 if unbiased else 0))).sqrt()`

File: tinygrad/tensor.py [hunk 3]
```diff
@@ -306,7 +306,9 @@ def min(self, axis=None, keepdim=False): return -((-self).max(axis=axis, keepdim
   def mean(self, axis=None, keepdim=False):
     out = self.sum(axis=axis, keepdim=keepdim)
     return out * (prod(out.shape)/prod(self.shape))
-
+  def std(self, axis=None, keepdim=False, unbiased=True):
+    square_sum = ((self - self.mean(axis=axis, keepdim=True)).square()).sum(axis=axis, keepdim=keepdim)
+    return (square_sum / ((prod(self.shape[:axis+1]) if axis is not None else prod(self.shape))-(1 if unbiased else 0))).sqrt()
```
Comment (geohot) [inline tinygrad/tensor.py:311]: prod(square_sum.shape)!

Comment (rabiaedayilmaz) [reply tinygrad/tensor.py:311]: Right, it does not pass. I see, I will try something else.

Comment (rabiaedayilmaz) [reply tinygrad/tensor.py]: My last commit `prod(square_sum[:axis])` passes tests, does it handle higher axes?

**Comment by geohot** [comment]
test axis=[1,2]?

**Comment by rabiaedayilmaz** [comment]
tested axis=[1, 2]

**Comment by rabiaedayilmaz** [comment]
No more silly mistakes, all is done.

File: tinygrad/tensor.py [hunk 4]
```diff
@@ -306,7 +306,10 @@ def min(self, axis=None, keepdim=False): return -((-self).max(axis=axis, keepdim
   def mean(self, axis=None, keepdim=False):
     out = self.sum(axis=axis, keepdim=keepdim)
     return out * (prod(out.shape)/prod(self.shape))
-
+  # TODO: unbiased True 0.01 difference with torch.std
```
Comment (geohot) [inline tinygrad/tensor.py:309]: So it's not ready to merge yet...

Comment (rabiaedayilmaz) [reply tinygrad/tensor.py:309]: Can I implement it without the unbiased option, like numpy, just for now @geohot ?

Comment (symlons) [reply tinygrad/tensor.py]: The 0.01 error seems to bee an issue with pytorch.

Comment (rabiaedayilmaz) [reply tinygrad/tensor.py]: That actually makes sense but how?

**Pull request merged**